File: player.ex
defmodule Player do
  defstruct name: "", level: 1, experience: 0, inventory: []

  def add_item(player, item) do
    %{player | inventory: [item | player.inventory]}
  end

  def remove_item(player, item) do
    %{player | inventory: Enum.filter(player.inventory, fn x -> x != item end)}
  end

  def gain_exp(player, points) do
    new_exp = player.experience + points
    new_level = if new_exp >= 100, do: player.level + 1, else: player.level
    final_exp = if new_exp >= 100, do: new_exp - 100, else: new_exp
    %{player | experience: final_exp, level: new_level}
  end
end

File: Player.hs
module Player where
import Data.List (delete)

data Player = Player {
    name :: String,
    level :: Int,
    experience :: Int,
    inventory :: [String]
} deriving Show

initPlayer :: String -> Player
initPlayer n = Player n 1 0 []

addItem :: Player -> String -> Player
addItem p item = p { inventory = item : inventory p }

removeItem :: Player -> String -> Player
removeItem p item = p { inventory = delete item (inventory p) }

gainExp :: Player -> Int -> Player
gainExp p points =
    let newExp = experience p + points
        newLevel = if newExp >= 100 then level p + 1 else level p
        finalExp = if newExp >= 100 then newExp - 100 else newExp
    in p { level = newLevel, experience = finalExp }
